### PR TITLE
Support RC2

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## RC1

```
$ docker run --rm -itv icingadb:/data -e ICINGADB_REDIS_HOST=172.17.0.3 -e ICINGADB_REDIS_PORT=6379 -e ICINGADB_MYSQL_HOST=172.17.0.2 -e ICINGADB_MYSQL_PORT=3306 -e ICINGADB_MYSQL_DATABASE=idb -e ICINGADB_MYSQL_USER=idb -e ICINGADB_MYSQL_PASSWORD=idb icinga/icingadb:rc1
DEBU[0000] translating env vars to .ini config
DEBU[0000] checking SQL database
DEBU[0000] importing schema into SQL database
DEBU[0001] starting actual daemon via exec(3)
INFO[0000] Connecting to Redis
```

## Upgrade and config int-to-string works

```
$ docker run --rm -itv icingadb:/data -e ICINGADB_REDIS_ADDRESS=172.17.0.3:6379 -e ICINGADB_REDIS_PASSWORD=123456 -e ICINGADB_DATABASE_HOST=172.17.0.2 -e ICINGADB_DATABASE_PORT=3306 -e ICINGADB_DATABASE_DATABASE=idb -e ICINGADB_DATABASE_USER=idb -e ICINGADB_DATABASE_PASSWORD=idb icinga/icingadb:rc2
2021-06-01T12:17:25.914Z	DEBUG	entrypoint/main.go:46	translating env vars to YaML config
2021-06-01T12:17:25.915Z	DEBUG	entrypoint/main.go:103	checking SQL database
2021-06-01T12:17:25.920Z	INFO	entrypoint/main.go:153	upgrading SQL database schema
2021-06-01T12:17:25.920Z	INFO	entrypoint/main.go:160	.. to v2
2021-06-01T12:17:26.914Z	DEBUG	entrypoint/main.go:92	starting actual daemon via exec(3)
2021-06-01T12:17:26.967Z	INFO	icingadb/main.go:51	Starting Icinga DB
```

File: entrypoint/main.go [hunk 1]
```diff
@@ -3,47 +3,86 @@
...
+			sectionCfg, ok := cfg[section]
+			if !ok {
+				sectionCfg = map[string]interface{}{}
+				cfg[section] = sectionCfg
+			}
+
+			rawValue := match[3]
+			var value interface{}
+
+			if isInt.MatchString(rawValue) {
```
Comment (julianbrost) [inline entrypoint/main.go]: Why this additional check? Shouldn't this already be covered by ParseInt returning an error?

File: entrypoint/main.go [hunk 2]
```diff
@@ -3,47 +3,86 @@
...
 			}
+
+			sectionCfg[strings.ToLower(match[2])] = value
 		}
 	}
 
-	if errST := cfg.SaveTo(config); errST != nil {
-		return errST
+	buf := &bytes.Buffer{}
+	buf.Write([]byte("---\n"))
```
Comment (julianbrost) [inline entrypoint/main.go]: Is there any need for this?

File: entrypoint/main.go [hunk 3]
```diff
@@ -52,55 +91,85 @@ func runDaemon() error {
 
 	log.Debug("starting actual daemon via exec(3)")
 
-	return syscall.Exec("/icingadb", []string{"/icingadb", "-config", config}, os.Environ())
+	return syscall.Exec("/icingadb", []string{"/icingadb", "-c", config, "--datadir", "/data"}, os.Environ())
```
Comment (julianbrost) [inline entrypoint/main.go]: Should consistently use short or long flags.

File: entrypoint/go.mod [hunk 4]
```diff
@@ -4,9 +4,6 @@ go 1.14
 
 require (
 	github.com/Icinga/go-libs v0.0.0-20200629140135-57268fcc7724
-	github.com/Icinga/icingadb v1.0.0-rc1
-	github.com/go-ini/ini v1.62.0
-	github.com/go-sql-driver/mysql v1.6.0
-	github.com/sirupsen/logrus v1.8.1
-	gopkg.in/ini.v1 v1.57.0 // indirect
+	github.com/icinga/icingadb v1.0.0-rc1.0.20210528122710-9fdcc839798a
```
Comment (julianbrost) [inline entrypoint/go.mod:7]: I don't like pinning any version here. This should better use `replace` to use the checkout of the icingadb repo that is used for the build.

File: entrypoint/main.go [hunk 5]
```diff
@@ -127,3 +196,54 @@ func dbHasSchema(db *sql.DB, dbName string) (bool, error) {
...
 	return len(res.([]one)) > 0, nil
 }
+
+func getDbSchemaVersion(db *sql.DB) (uint64, error) {
+	type icingadbSchema struct {
+		Version uint64
+	}
+
+	rows, errQr := db.Query(
+		"SELECT version FROM icingadb_schema ORDER BY `timestamp` DESC LIMIT 1",
```
Comment (julianbrost) [inline entrypoint/main.go]: `SELECT MAX(version) FROM icingadb_schema;`?

**Review by julianbrost — CHANGES_REQUESTED** [review]
The `icingadb` binary currently takes a `--datadir` parameter, but it's currently not used anywhere and I don't know of any plans to use it (there was one, that's why the parameter exists, but that plan was discarded). So do we really want to annoy users with configuring a data volume that's not used? I'd probably rather remove the parameter. @lippserd What do you think?

Also, didn't we agree on not doing automatic schema upgrades for now? At least the current upgrade file is not safe for being applied twice in parallel and there seem to be no safeguards here that would mitigate this.

File: entrypoint/main.go (hunk 3 above)
Comment (Al2Klimov) [reply entrypoint/main.go]: Why? Not this repo's problem if an upstream flag doesn’t have a short version.

File: entrypoint/go.mod (hunk 4 above)
Comment (Al2Klimov) [reply entrypoint/go.mod:7]: Once we have a stable version, @dependabot will take care of this.

File: entrypoint/main.go (hunk 5 above)
Comment (Al2Klimov) [reply entrypoint/main.go]: Downgrades?

**Comment by Al2Klimov** [comment]
1. It will work w/o a volume as well. But don’t tell anyone. 😉
2. Safeguard: #27

File: entrypoint/main.go (hunk 3 above)
Comment (julianbrost) [reply entrypoint/main.go]: There's a long version for both. Why not go with that?

Comment (Al2Klimov) [reply entrypoint/main.go]: Oh, c'mon. The next time I see you typing a tar command with many short flags and one long one, I'll say this to you.

File: README.md [hunk 6]
```diff
@@ -26,31 +26,26 @@ docker run --rm -d \
...
 Each environment variable corresponds to a configuration option of Icinga DB.
-E.g. `ICINGADB_REDIS_HOST=2001:db8::192.0.2.18` means:
+E.g. `ICINGADB_REDIS_ADDRESS=2001:db8::192.0.2.18` means:
 
-```ini
-[redis]
-host = "2001:db8::192.0.2.18"
+```yaml
+redis:
+  address: 2001:db8::192.0.2.18
```
Comment (julianbrost) [inline README.md]: Does this work or should this be something like `[2001:db8::192.0.2.18]:6379`.

Also why use this mixed notation for something else than `::ffff:0:0/96` and NAT64 addresses? It only adds confusion.

File: entrypoint/main.go [hunk 7]
```diff
@@ -23,27 +23,53 @@ var myEnv = regexp.MustCompile(`(?s)\AICINGADB_(\w+?)_(\w+)=(.*)\z`)
...
 func main() {
-	log.SetOutput(os.Stdout)
-	log.SetLevel(log.TraceLevel)
-	log.Fatal(runDaemon())
+	log.Fatal("failed", zap.Error(runDaemon()))
 }
 
 func runDaemon() error {
-	log.Debug("translating env vars to .ini config")
+	log.Debug("translating env vars to YaML config")
```
Comment (julianbrost) [inline entrypoint/main.go:40]: I haven't seen this capitalization of YAML before. https://yaml.org/ also uses all uppercase.

**Review by julianbrost — CHANGES_REQUESTED** [review]
Tested it and it works in general.

However, when running the container without any environment variables set, it just segfaults. This could be handled in a more graceful way.
```
2021-06-08T13:47:17.710Z	DEBUG	entrypoint/main.go:40	translating env vars to YaML config
2021-06-08T13:47:17.710Z	DEBUG	entrypoint/main.go:85	checking SQL database
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x6bfdc2]

goroutine 1 [running]:
github.com/icinga/icingadb/pkg/config.(*Database).Open(0x0, 0xc00000e178, 0xc0001a8800, 0x0, 0x0)
	/go/pkg/mod/github.com/icinga/icingadb@v1.0.0-rc1.0.20210528122710-9fdcc839798a/pkg/config/database.go:37 +0x82
main.initDb(0x0, 0x0)
	/entrypoint/main.go:92 +0x16a
main.runDaemon(0x7504f0, 0x0)
	/entrypoint/main.go:75 +0x165
main.main()
	/entrypoint/main.go:36 +0x37
```

On `/data`: The flag exists because there once was the plan to persist the instance ID somewhere. However, we decided that it would be nicer to keep Icinga DB stateless and thus implemented the instance ID differently. I think it's strange to prepare something, that we currently don't use, have no plan for using it, and don't want to do if we don't have to (at least in my opinion). @lippserd Any opinion on this, as you added the `--datadir` flag in the first place.

**Comment by lippserd** [comment]
> @lippserd Any opinion on this, as you added the `--datadir` flag in the first place.

We don't need `--datadir` anymore as we will be removing it from the Icinga DB.

File: entrypoint/main.go (hunk 7 above)
Comment (julianbrost) [reply entrypoint/main.go:40]: What about this?

**Review by julianbrost — COMMENTED** [review]
What about my remaining two points (the one below and that `docker run --rm -it icinga/icingadb` panics)?

We can probably merge this as is, but that should still be addressed at some point.

**Review by julianbrost — APPROVED** [review]
Merging as-is to get back working images, creating an issue for the remaining feedback

**Pull request merged**